<?xml version="1.0"?>
<launch>
	<!-- Launch file for PX4 and IBVS -->
	<!-- JSW May 2018 -->

    <!--  -->
    <!-- ARGS -->
    <!--  -->

    <!-- rosbag args -->
    <arg name="record_rosbag" default="false" />
    <arg name="bag_name" default="ibvs" />

    <!-- Point Grey Camera Args -->
    <arg name="frame_rate" value="20" />
    <arg name="calibrated" default="true" />

    <!-- ArUco args -->
    <arg name="show_aruco_detections" default="false" />
	
    
    <!--  -->
    <!-- PX4 -->
    <!--  -->

    <!-- Launch PX4 pixhawk (taken from mavros px4.launch) -->
	<arg name="fcu_url" default="/dev/ttyUSB0:921600" />
	<arg name="gcs_url" default="" />
	<arg name="tgt_system" default="1" />
	<arg name="tgt_component" default="1" />
	<arg name="log_output" default="screen" />

	<include file="$(find mavros)/launch/node.launch">
	    <arg name="pluginlists_yaml" value="$(find mavros)/launch/px4_pluginlists.yaml" />
		<arg name="config_yaml" value="$(find mavros)/launch/px4_config.yaml" />

		<arg name="fcu_url" value="$(arg fcu_url)" />
		<arg name="gcs_url" value="$(arg gcs_url)" />
		<arg name="tgt_system" value="$(arg tgt_system)" />
		<arg name="tgt_component" value="$(arg tgt_component)" />
		<arg name="log_output" value="$(arg log_output)" />
	</include>


	<!--  -->
    <!-- MAVROS NED -->
    <!--  -->

    <!-- Node to convert mavros data to estimate data -->
	<node name="mavros_ned" pkg="ibvs_sim" type="mavros_ned" output="screen" />


    <!--  -->
    <!-- IBVS Control -->
    <!--  -->

    <rosparam command="load" file="$(find ibvs_sim)/params/ibvs.yaml"/>

    <!-- p_des Publisher -->
    <node name="p_des_mapper" pkg="ibvs_sim" type="p_des_mapper.py" output="screen" />
    
    <!-- IBVS Outer -->
    <node name="ibvs" pkg="ibvs_sim" type="ibvs.py" output="screen">
        <remap from="/aruco/marker_corners" to="/aruco/marker_corners_outer"/>
        <remap from="/aruco/distance" to="/aruco/distance_outer"/>
        <remap from="/quadcopter/camera/camera_info" to="/camera_info" />
    </node>

    <!-- IBVS Inner -->
    <node name="ibvs_inner" pkg="ibvs_sim" type="ibvs.py" output="screen">
        <remap from="/aruco/marker_corners" to="/aruco/marker_corners_inner"/>
        <remap from="/aruco/distance" to="/aruco/distance_inner"/>
        <remap from="/ibvs/vel_cmd" to="/ibvs_inner/vel_cmd"/>
        <remap from="/quadcopter/camera/camera_info" to="/camera_info" />
    </node>


    <!-- STOPED HERE, NEED TO setup the state machine ArUco and Camera below -->
    <!-- STOPED HERE, NEED TO setup the state machine ArUco and Camera below -->
    <!-- STOPED HERE, NEED TO setup the state machine ArUco and Camera below -->
    <!-- STOPED HERE, NEED TO setup the state machine ArUco and Camera below -->

    <!-- IBVS State Machine -->
    <node name="ibvs_state_machine" pkg="ibvs_sim" type="ibvs_sm_mavros.py" output="screen" >
    	<remap from="/quadcopter/camera/camera_info" to="/camera_info" />

    	<param name="u_max" value="0.2" />
    	<param name="v_max" value="0.2" />
    	<param name="w_max" value="0.4" />

        <param name="u_max_inner" value="0.2" />
        <param name="v_max_inner" value="0.2" />
        <param name="w_max_inner" value="0.4" />
        

    	<param name="wp_N" value="0.0" />
    	<param name="wp_E" value="0.0" />
    	<param name="wp_D" value="-10.0" />

        <param name="count_outer" value="50" />
        <param name="count_inner" value="20" />
    </node>

    <!-- ArUco Localization -->
    <node pkg="aruco_localization" type="aruco_localization" name="aruco" output="screen">
        <param name="show_output_video" value="$(arg show_aruco_detections)" />
        <param name="markermap_config" value="$(find ibvs_sim)/params/nested148_65.yaml" />
        <param name="marker_size" value="0.7071" />
        <param name="marker_size_inner" value="0.127" />

        <param name="id_outer" value="148" />
        <param name="id_inner" value="65" />
        <param name="debug_save_input_frames" value="false" />
        <param name="debug_save_output_frames" value="false" />
        <param name="debug_image_path" value="/tmp/arucoimages" />

        <remap from="input_image" to="image_raw" />
        <remap from="output_image" to="aruco/image" />
    </node>

    <!-- Point Grey Camera -->
    <node name="pointgrey_camera" pkg="pointgrey_camera_driver" type="camera_node">

  	    <param name="camera_info_url" if="$(arg calibrated)"
            value="file://$(find aruco_localization)/camera_info/chameleon3_LLNL.yaml" />
        <param name="frame_rate" value="$(arg frame_rate)" />
        <param name="auto_shutter" value="true" />
        <!-- <param name="shutter_speed" value="0.03" /> -->
        <param name="format7_color_coding" value="mono8" />
        <param name="auto_white_balance" value="true" />
    </node>

    <!-- rosbag record -->
    <group if="$(arg record_rosbag)">
        <node name="record" pkg="rosbag" type="record" args="/ibvs/uv_bar_lf_inner /ibvs/uv_bar_lf /aruco/marker_corners_outer /aruco/marker_corners_inner /camera_info /image_raw /aruco/estimate -o $(arg bag_name)" />
    </group>



</launch>
